src/librustc_typeck/check/method/probe.rs

```diff
@@ -1068,10 +1079,19 @@ impl<'a,'tcx> ProbeContext<'a,'tcx> {
                     debug!("impl_obligations={}", obligations.repr(self.tcx()));
 
                     // Evaluate those obligations to see if they might possibly hold.
-                    obligations.iter()
-                        .chain(norm_obligations.iter()).chain(ref_obligations.iter())
-                        .all(|o| selcx.evaluate_obligation(o))
-
+                    let mut all_true = true;
+                    for o in obligations.iter()
+                        .chain(norm_obligations.iter())
+                        .chain(ref_obligations.iter()) {
+                        if !selcx.evaluate_obligation(o) {
+                            all_true = false;
+                            // FIXME: is this the only case we care about?
```

Comment:
I would say just limit it to `Predicate::Trait`. The others kinds are typically "evaluated" to "true" anyway, and we can always extend it. (So feel free to remove the FIXME.)